packages/hub/src/lib/oauth-login.ts

```diff
@@ -0,0 +1,119 @@
+	/**
+	 * OAuth scope, a list of space separate scopes.
+	 *
+	 * For static Spaces, you can omit this and it will be loaded from the Space config, as long as `hf_oauth: true` is present in the README.md's metadata.
+	 * For other Spaces, it is available to the backend in the OAUTH_SCOPES environment variable, as long as `hf_oauth: true` is present in the README.md's metadata.
+	 *
+	 * Defaults to "openid profile".
+	 *
+	 * You can also create a Developer Application at https://huggingface.co/settings/connected-applications and use its scopes.
+	 *
+	 * See https://huggingface.co/docs/hub/oauth for a list of available scopes.
+	 */
+	scopes?: string;
+	/**
+	 * Redirect URI, defaults to the current URL.
+	 *
+	 * For Spaces, any URL within the Space is allowed.
+	 *
+	 * For Developer Applications, you can add any URL you want to the list of allowed redirect URIs at https://huggingface.co/settings/connected-applications.
+	 */
+	redirectUri?: string;
+	/**
+	 * State to pass to the OAuth provider, which will be returned in the call to `oauthLogin` after the redirect.
+	 */
+	state?: string;
+}): Promise<void> {
+	if (typeof window === "undefined") {
+		throw new Error("oauthLogin is only available in the browser");
+	}
+
+	const hubUrl = opts?.hubUrl || HUB_URL;
+	const openidConfigUrl = `${new URL(hubUrl).origin}/.well-known/openid-configuration`;
+	const openidConfigRes = await fetch(openidConfigUrl, {
+		headers: {
+			Accept: "application/json",
+		},
+	});
+
+	if (!openidConfigRes.ok) {
+		throw await createApiError(openidConfigRes);
+	}
+
+	const opendidConfig: {
+		authorization_endpoint: string;
+		token_endpoint: string;
+		userinfo_endpoint: string;
+	} = await openidConfigRes.json();
+
+	const newNonce = crypto.randomUUID();
+	// Two random UUIDs concatenated together, because min length is 43 and max length is 128
+	const newCodeVerifier = crypto.randomUUID() + crypto.randomUUID();
+
+	localStorage.setItem("huggingface.co:oauth:nonce", newNonce);
+	localStorage.setItem("huggingface.co:oauth:code_verifier", newCodeVerifier);
+
+	const redirectUri = opts?.redirectUri || window.location.href;
+	const state = JSON.stringify({
+		nonce: newNonce,
+		redirectUri,
+		state: opts?.state,
+	});
+
+	// @ts-expect-error window.huggingface is defined inside static Spaces.
+	const variables: Record<string, string> | null = window?.huggingface?.variables ?? null;
+
+	const clientId = opts?.clientId || variables?.OAUTH_CLIENT_ID;
+
+	if (!clientId) {
+		if (variables) {
+			throw new Error("Missing clientId, please add hf_oauth: true to the README.md's metadata in your static Space");
+		}
+		throw new Error("Missing clientId");
+	}
+
+	const challenge = base64FromBytes(
+		new Uint8Array(await globalThis.crypto.subtle.digest("SHA-256", new TextEncoder().encode(newCodeVerifier)))
+	)
+		.replace(/[+]/g, "-")
+		.replace(/[/]/g, "_")
+		.replace(/=/g, "");
+
+	window.location.href = `${opendidConfig.authorization_endpoint}?${new URLSearchParams({
```

Comment:
would it make sense to have an option to use `window.open`  and optionally open the oauth login as an external window or tab?